src/core/lib/surface/server.cc

```diff
@@ -1433,30 +1448,40 @@ static grpc_call_error queue_call_request(grpc_server* server, size_t cq_idx,
       rm = &rc->data.registered.method->matcher;
       break;
   }
-  if (gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link)) {
-    /* this was the first queued request: we need to lock and start
-       matching calls */
-    gpr_mu_lock(&server->mu_call);
-    while ((calld = rm->pending_head) != nullptr) {
-      rc = reinterpret_cast<requested_call*>(
-          gpr_locked_mpscq_pop(&rm->requests_per_cq[cq_idx]));
-      if (rc == nullptr) break;
-      rm->pending_head = calld->pending_next;
-      gpr_mu_unlock(&server->mu_call);
-      if (!gpr_atm_full_cas(&calld->state, PENDING, ACTIVATED)) {
-        // Zombied Call
-        GRPC_CLOSURE_INIT(
-            &calld->kill_zombie_closure, kill_zombie,
-            grpc_call_stack_element(grpc_call_get_call_stack(calld->call), 0),
-            grpc_schedule_on_exec_ctx);
-        GRPC_CLOSURE_SCHED(&calld->kill_zombie_closure, GRPC_ERROR_NONE);
-      } else {
-        publish_call(server, calld, cq_idx, rc);
-      }
-      gpr_mu_lock(&server->mu_call);
-    }
+
+  // Fast path: if there is no pending request to be processed, immediately
+  // return.
+  if (!gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link) ||
+      // Note: pending_head is written using order_release in publish_new_rpc().
+      //       Here we are reading the pointer without holding the sever call
+      //       mutex. The reset of stores and loads in this function are under
+      //       server call mutex and can use order_relaxed.
+      rm->pending_head.load(std::memory_order_acquire) == nullptr) {
```

Comment:
This acquire is almost certainly unnecessary.

In any context where this load matches with the store above (and it can't match with any other store because that is the only one that is release) we immediately acquire the lock below and then double-check.

So the only time this could possibly matter is if we do observe null, at which point we don't have a matching release.

Reply:
Thank you. Yes, agreed. I was going too pedantic. Also, there are already barriers around this.